lib/spack/spack/test/cmd/create.py

```diff
@@ -40,18 +24,87 @@ def parser():
     return prs
 
 
-def test_create_template(parser, cmd_create_repo):
+@pytest.mark.parametrize('args,name_index,expected', [
+    (['test-package'], 0, [r'TestPackage(Package)', r'def install(self']),
+    (['-n', 'test-named-package', 'file://example.tar.gz'], 1,
+     [r'TestNamedPackage(Package)', r'def install(self']),
+    (['file://example.tar.gz'], -1,
+     [r'Example(Package)', r'def install(self']),
+    (['-t', 'bundle', 'test-bundle'], 2, [r'TestBundle(BundlePackage)'])
+])
+def test_create_template(parser, mock_test_repo, args, name_index, expected):
     """Test template creation."""
-    repo, repodir = cmd_create_repo
+    repo, repodir = mock_test_repo
 
-    name = 'test-package'
-    args = parser.parse_args(['--skip-editor', name])
-    spack.cmd.create.create(parser, args)
+    constr_args = parser.parse_args(['--skip-editor'] + args)
+    spack.cmd.create.create(parser, constr_args)
 
-    filename = repo.filename_for_package_name(name)
+    pkg_name = args[name_index] if name_index > -1 else 'example'
+    filename = repo.filename_for_package_name(pkg_name)
     assert os.path.exists(filename)
 
     with open(filename, 'r') as package_file:
         content = ' '.join(package_file.readlines())
-        for entry in [r'TestPackage(Package)', r'def install(self']:
-            assert content.find(entry) > -1
+        for entry in expected:
+            assert entry in content
+
+
+@pytest.mark.parametrize('name,expected', [
+    (' ', 'name must be provided'),
+    ('bad#name', 'name can only contain'),
+])
+def test_create_template_bad_name(parser, mock_test_repo, name, expected):
+    """Test template creation with bad name options."""
+    constr_args = parser.parse_args(['--skip-editor', '-n', name])
+    with pytest.raises(SystemExit, matches=expected):
+        spack.cmd.create.create(parser, constr_args)
+
+
+def test_build_system_guesser_no_stage(parser):
+    """Test build system guesser when stage not provided."""
+    guesser = spack.cmd.create.BuildSystemGuesser()
+
+    # Ensure get the expected build system
+    with pytest.raises(AttributeError,
+                       matches="'NoneType' object has no attribute"):
+        guesser(None, '/the/url/does/not/matter')
+
+
+def test_build_system_guesser_octave(parser):
+    """Test build system guesser for the octave URL special case."""
```

Comment:
Could you explain what makes it a special case? It seems that normally `guesser(None)` is supposed to fail so I assume it has something to do with that.

Reply:
The difference is a common URL versus guessing based on the contents of the archive file.

Expanded the comment.